server/grpc_service.go

```diff
@@ -59,6 +59,28 @@ var (
 	ErrNotStarted = status.Errorf(codes.Unavailable, "server not started")
 )
 
+type Header interface {
+	GetHeader() *pdpb.RequestHeader
+}
+
+type forwardFn func(ctx context.Context, client *grpc.ClientConn) (interface{}, error)
+
+func (s *GrpcServer) forwardRequest(ctx context.Context, header Header, fn forwardFn) (rsp interface{}, err error) {
+	forwardedHost := getForwardedHost(ctx)
+	if !s.isLocalRequest(forwardedHost) {
+		client, err := s.getDelegateClient(ctx, forwardedHost)
+		if err != nil {
+			return nil, err
+		}
+		ctx = grpcutil.ResetForwardContext(ctx)
+		return fn(ctx, client)
+	}
+	if err := s.validateRequest(header.GetHeader()); err != nil {
```

Comment:
I prefer not to contain it in `forwardRequest`.

Reply:
I think the forward request should validate the request also .

Reply:
Yes, but the function name should be changed. For now, it sounds like only forward requests need to be validated.